### PR TITLE
Feature/skip take paggination api

### PR DESCRIPTION
**Pull request by Rabosa616**
Add Paging support using the Skip() Take() 
Related:

- #149

### CODE REVIEW

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs [hunk 1]
```diff
@@ -357,6 +357,37 @@ await container.ReadItemAsync<TItem>(id, partitionKey, cancellationToken: cancel
...
+            Response<int> countResponse = await query.CountAsync(cancellationToken);
+            query = query.Skip(pageSize * (pageNumber - 1))
+                         .Take(pageSize);
+
+            using FeedIterator<TItem> iterator = query.ToFeedIterator();
+
+            FeedResponse<TItem> next = await iterator.ReadNextAsync(cancellationToken)
+                                                     .ConfigureAwait(false);
+
+            List<TItem> results = new();
```
Comment (mumby0168) [inline Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:377]: This looks good there is some cases where cosmos may not return all the results in a single call to `ReadNextAsync()` so you need to do this in a `while` loop as current paging example does. I'll find the link to the docs shortly

File: Microsoft.Azure.CosmosRepository/src/Paging/IPage.cs [hunk 2]
```diff
@@ -17,6 +17,16 @@ public interface IPage<out T> where T : IItem
         /// </summary>
         int Total { get; }
 
+        /// <summary>
```
Comment (mumby0168) [inline Microsoft.Azure.CosmosRepository/src/Paging/IPage.cs:20]: How do you see these extra properties work when using the existing API? All these properites would not be populated, could we maybe derive another interface from this one with a different name that offers these extra properties? What do you think?

File: Microsoft.Azure.CosmosRepository/src/Paging/Page.cs [hunk 3]
```diff
@@ -40,5 +77,33 @@ internal Page(int total, int size, IReadOnlyList<T> items, double charge, string
...
+        public int NextPageNumber => GetNextPageNumber();
+
+        private int GetNextPageNumber()
+        {
+            return HasNextPage && PageNumber.HasValue ? PageNumber.Value + 1 : TotalPages;
+        }
+
+        private int GetTotalPages()
+        {
+            int result = (int)Math.Ceiling(Total / (double)Size);
```
Comment (mumby0168) [inline Microsoft.Azure.CosmosRepository/src/Paging/Page.cs:100]: Could we add some tests for these methods, if you don't mind :)

**Comment by mumby0168** [comment]
Looks a great start, just a few comments/questions. Thanks for your great work so far 😃

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs (hunk 1 above)
Comment (mumby0168) [reply Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:377]: See this comment here as part of the initial paging PR, it includes the link I was referring to

https://github.com/IEvangelist/azure-cosmos-dotnet-repository/pull/142#discussion_r763042915

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs [hunk 4]
```diff
@@ -324,12 +327,54 @@ await container.ReadItemAsync<TItem>(id, partitionKey, cancellationToken: cancel
...
+            return new Page<TItem>(
+                countResponse.Resource,
+                pageSize,
+                result.Items.AsReadOnly(),
+                result.Charge + countResponse.RequestCharge,
+                result.ContinuationToken);
+        }
+
+        /// <inheritdoc/>
+        public async ValueTask<IPageExtended<TItem>> PageAsync(Expression<Func<TItem, bool>> predicate = null, int pageNumber = 1, int pageSize = 25, CancellationToken cancellationToken = default)
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:341]:
```suggestion
        public async ValueTask<IPageExtended<TItem>> PageAsync(
            Expression<Func<TItem, bool>> predicate = null,
            int pageNumber = 1,
            int pageSize = 25,
            CancellationToken cancellationToken = default)
```

Style nit

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs [hunk 5]
```diff
@@ -324,12 +327,54 @@ await container.ReadItemAsync<TItem>(id, partitionKey, cancellationToken: cancel
...
+
+        static void TryLogDebugDetails(ILogger logger, Func<string> getMessage)
+        {
+            if (logger?.IsEnabled(LogLevel.Debug) ?? false)
+            {
+                logger.LogDebug(getMessage());
+            }
+        }
 
+        static async Task<(List<TItem> items, double charge, string continuationToken)> GetAllItems(IQueryable<TItem> query, int pageSize, CancellationToken cancellationToken = default)
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:371]:
```suggestion
        static async Task<(List<TItem> items, double charge, string continuationToken)> GetAllItemsAsync(
            IQueryable<TItem> query,
            int pageSize,
            CancellationToken cancellationToken = default)
```

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs [hunk 6]
```diff
@@ -324,12 +327,54 @@ await container.ReadItemAsync<TItem>(id, partitionKey, cancellationToken: cancel
...
+            Container container = await _containerProvider.GetContainerAsync().ConfigureAwait(false);
+
+            IQueryable<TItem> query = container.GetItemLinqQueryable<TItem>()
+                                               .Where(_repositoryExpressionProvider.Build(predicate));
+
+            Response<int> countResponse = await query.CountAsync(cancellationToken);
+            query = query.Skip(pageSize * (pageNumber - 1))
+                         .Take(pageSize);
+
+            (List<TItem> Items, double Charge, string ContinuationToken) result = await GetAllItems(query, pageSize, cancellationToken);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:352]:
```suggestion
            (List<TItem> Items, double Charge, string ContinuationToken) result =
                await GetAllItemsAsync(query, pageSize, cancellationToken);
```

File: Microsoft.Azure.CosmosRepository/src/Paging/IPageExtended.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+#nullable enable
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.CosmosRepository.Paging
+{
+    /// <summary>
+    /// Represents a page of data from a cosmos query
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public interface IPageExtended<out T>:IPage<T> where T : IItem
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Paging/IPageExtended.cs:13]:
```suggestion
    public interface IPageExtended<out T> : IPage<T> where T : IItem
```

File: Microsoft.Azure.CosmosRepository/src/Paging/PageExtended.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+// Licensed under the MIT License.
+
+#nullable enable
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.CosmosRepository.Paging
+{
+    /// <inheritdoc/>
+    public class PageExtended<T> : Page<T>, IPageExtended<T> where T : IItem
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Paging/PageExtended.cs:11]: I'm not sure I like the name `PageExtended`, it's not really descriptive. What do you think about `PageQueryResult` or something like that? The interface would have to change too.

File: Microsoft.Azure.CosmosRepository/src/Paging/PageExtended.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        /// <inheritdoc />
+        public int PreviousPageNumber => GetPreviousPageNumber();
+
+        /// <inheritdoc />
+        public int NextPageNumber => GetNextPageNumber();
+
+        private int GetNextPageNumber()
+        {
+            return HasNextPage && PageNumber.HasValue ? PageNumber.Value + 1 : TotalPages;
+        }
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Paging/PageExtended.cs:72]:
```suggestion
        private int GetNextPageNumber() =>
            HasNextPage && PageNumber.HasValue ? PageNumber.Value + 1 : TotalPages;
```

File: Microsoft.Azure.CosmosRepository/src/Paging/PageExtended.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+        private int GetTotalPages()
+        {
+            return (int)Math.Abs(Math.Ceiling(Total / (double)Size));
+        }
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Paging/PageExtended.cs:77]:
```suggestion
        private int GetTotalPages() =>
            (int)Math.Abs(Math.Ceiling(Total / (double)Size));
```

File: Microsoft.Azure.CosmosRepository/src/Paging/PageExtended.cs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+        private int GetPreviousPageNumber()
+        {
+            return HasPreviousPage && PageNumber.HasValue ? PageNumber.Value - 1 : 1;
+        }
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Paging/PageExtended.cs:82]:
```suggestion
        private int GetPreviousPageNumber() =>
            HasPreviousPage && PageNumber.HasValue ? PageNumber.Value - 1 : 1;
```

Prefer expression-bodied members when possible.

File: README.md [hunk 12]
```diff
@@ -416,7 +416,7 @@ Visit the `Microsoft.Azure.CosmosRepository.Samples` [directory](https://github.
 - [Azure Functions](https://github.com/IEvangelist/azure-cosmos-dotnet-repository/tree/main/Microsoft.Azure.CosmosRepository.Samples/AzureFunctionTier)
 - [Services](https://github.com/IEvangelist/azure-cosmos-dotnet-repository/tree/main/Microsoft.Azure.CosmosRepository.Samples/ServiceTier)
 - [Controllers (web apps)](https://github.com/IEvangelist/azure-cosmos-dotnet-repository/tree/main/Microsoft.Azure.CosmosRepository.Samples/WebTier)
-
+- [Paging](https://github.com/IEvangelist/azure-cosmos-dotnet-repository/tree/main/Microsoft.Azure.CosmosRepository.Samples/Paging)
```
Comment (IEvangelist) [inline README.md:419]: Nice to see another example in the README. 😁

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs [hunk 13]
```diff
@@ -324,12 +327,54 @@ await container.ReadItemAsync<TItem>(id, partitionKey, cancellationToken: cancel
                 .Where(_repositoryExpressionProvider.Build(predicate));
 
             Response<int> countResponse = await query.CountAsync(cancellationToken);
-            using FeedIterator<TItem> iterator = query.ToFeedIterator();
+            (List<TItem> Items, double Charge, string ContinuationToken) result = await GetAllItems(query, pageSize, cancellationToken);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:330]:
```suggestion
            (List<TItem> Items, double Charge, string ContinuationToken) result =
                await GetAllItemsAsync(query, pageSize, cancellationToken);
```

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs [hunk 14]
```diff
@@ -2,14 +2,17 @@
 // Licensed under the MIT License.
 
 using System;
+using System.Collections;
 using System.Collections.Generic;
 using System.Linq;
 using System.Linq.Expressions;
 using System.Net;
+using System.Runtime.InteropServices;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:10]: Wait, why is this namespace needed?

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs [hunk 15]
```diff
@@ -2,14 +2,17 @@
...
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Azure.Cosmos;
 using Microsoft.Azure.Cosmos.Linq;
+using Microsoft.Azure.Cosmos.Serialization.HybridRow;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:15]: Same question, why is this namespace needed?

File: Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs [hunk 16]
```diff
@@ -324,12 +327,54 @@ await container.ReadItemAsync<TItem>(id, partitionKey, cancellationToken: cancel
...
+
+        /// <inheritdoc/>
+        public async ValueTask<IPageExtended<TItem>> PageAsync(Expression<Func<TItem, bool>> predicate = null, int pageNumber = 1, int pageSize = 25, CancellationToken cancellationToken = default)
+        {
+            Container container = await _containerProvider.GetContainerAsync().ConfigureAwait(false);
+
+            IQueryable<TItem> query = container.GetItemLinqQueryable<TItem>()
+                                               .Where(_repositoryExpressionProvider.Build(predicate));
+
+            Response<int> countResponse = await query.CountAsync(cancellationToken);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/DefaultRepository.cs:348]:
```suggestion
            Response<int> countResponse =
                await query.CountAsync(cancellationToken).ConfigureAwait(false);
```

File: Microsoft.Azure.CosmosRepository/src/IRepository.cs [hunk 17]
```diff
@@ -242,5 +242,21 @@ public interface IRepository<TItem> where TItem : IItem
...
+        /// <summary>
+        /// Offers a load more paging implementation for infinite scroll scenarios.
+        /// Allows for efficient paging making use of cosmos DBs continuation tokens, making this implementation cost effective.
+        /// </summary>
+        /// <param name="predicate">A filter criteria for the paging operation, if null it will get all <see cref="IItem"/>s</param>
+        /// <param name="pageNumber">The page number to return from cosmos db.</param>
+        /// <param name="pageSize">The size of the page to return from cosmos db.</param>
+        /// <param name="cancellationToken">The cancellation token to use when making asynchronous operations.</param>
+        /// <returns>A <see cref="IPageExtended{T}"/> of <see cref="IItem"/>s</returns>
+        /// <remarks>This method makes use of cosmos dbs continuation tokens for efficient, cost effective paging utilising low RUs</remarks>
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/IRepository.cs:255]:
```suggestion
        /// <remarks>This method makes use of Cosmos DB's continuation tokens for efficient, cost effective paging utilizing low RUs</remarks>

```

**Review by IEvangelist — COMMENTED** [review]
This is looking nice, well done and thank you. I have some questions, suggestions, and nits.

**Comment by IEvangelist** [comment]
Hey @all-contributors, please add @Rabosa616 for code, tests, and docs

**Comment by IEvangelist** [comment]
Hey @all-contributors please add @Rabosa616 for code, tests, and docs.

File: Microsoft.Azure.CosmosRepository/src/Paging/PageQueryResult.cs [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    public class PageQueryResult<T> : Page<T>, IPageQueryResult<T> where T : IItem
+    {
+        /// <summary>
+        /// Creates a page.
+        /// </summary>
+        /// <param name="total"></param>
+        /// <param name="size"></param>
+        /// <param name="items"></param>
+        /// <param name="charge"></param>
+        /// <param name="continuation"></param>
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Paging/PageQueryResult.cs:20]:
```suggestion
        /// <param name="total">The total.</param>
        /// <param name="size">The size.</param>
        /// <param name="items">The items.</param>
        /// <param name="charge">The charge.</param>
        /// <param name="continuation">The continuation.</param>
```

**Review by IEvangelist — DISMISSED** [review]

**Comment by IEvangelist** [comment]
I'll let @mumby0168 merge this, thank you - looks good.

**Comment by mumby0168** [comment]
> I'll let @mumby0168 merge this, thank you - looks good.

Will sort this tomorrow 👍🏼

**Review by mumby0168 — DISMISSED** [review]
Looks really good! Thanks again

File: Microsoft.Azure.CosmosRepository.Samples/Paging/Program.cs [hunk 19]
```diff
@@ -23,12 +23,29 @@
 
 await SeedAsync();
 await BasicScrollingAsync();
+await BasicPageAsync();
+
+async Task BasicPageAsync()
+{
+    double totalCharge = 0;
+    IPageQueryResult<Person> page = await repository.PageAsync(pageNumber:1,pageSize: 25);
```
Comment (mumby0168) [inline Microsoft.Azure.CosmosRepository.Samples/Paging/Program.cs:31]:
```suggestion
    IPageQueryResult<Person> page = await repository.PageAsync(pageNumber: 1,pageSize: 25);
```

**Review by mumby0168 — APPROVED** [review]

**Pull request merged**